src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexThresholds.cs

```diff
@@ -22,19 +22,24 @@ internal static class SymbolicRegexThresholds
         /// processing starts out in DFA mode, even if we've previously triggered NFA mode for the same regex.
         /// We switch over into NFA mode the first time a given traversal (match operation) results in us needing
         /// to create a new node and the graph is already or newly beyond this threshold.
+        /// This limit is chosen due to memory usage constraints, the worst possible allocation is currently approx. 50 MB;
+        /// There is some motivation to make this configurable, as it can exchange upfront costs with potentially
+        /// significant search-time performance gains.
         /// </remarks>
-        internal const int NfaThreshold = 10_000;
+        internal const int NfaThreshold = 25_000;
```

Comment:
Here we were testing with an intentionally very expensive pattern within the raised safe size threshold to see how much memory it allocates. It'd be best to tune this on some large patterns from RealWorldRegexes.json to see how it affects performance since the switch from DFA to NFA has a steep performance drop.

Is this 50MB worst case ok or should we go lower? Is it possible to make this configurable?

Reply:
> Is this 50MB worst case ok or should we go lower?

Can you elaborate on this? 50MBs of what? Are you saying 10K DFA nodes consume 20MB of managed memory and increasing it to 25K DFA nodes consumes 50MB of managed memory?

Reply:
I was wondering if 50MBs of managed memory is a reasonable upper limit for the engine before falling back to NFA mode (which has negligible memory use).

Although i'll still see if this memory limit actually gets enforced as well, because i think we're currently measuring the wrong thing, this NfaThreshold is measuring the number of DFA states which can vary a lot in memory use, we should be counting the amount of SymbolicRegexNode instances the builder has created instead.

Reply:
I just had a longer skype meeting with Ian about this. I agree. The total number of **SymbolicRegexNode** objects overapproximates the total sum of both what would correspond NFA states and the count of the number of DFA states (whose corresponding nodes consist of unions/alternations of NFA states as "powerset" states). The total number of such nodes very accurately reflects the memory usage. If the total reaches a **NodeCountThreshold** then the algorithm should switch to NFA mode. NodeCountThreshold value should correspond, in nr of nodes, to upper memory limit allowed, say 50MB divided by how many bytes each Node occupies, e.g., if this is say 500bytes per node (I believe it is considerably less) then the NodeCountThreshold could be set to 100k. 
This limit is much more accurate as it is direcltly related to memory usage and essentially eliminates the need for use of NfaThreshold  that just was an upper limit for DFA states not taking into account the total accumulated memory size,

Reply:
The typical bad case is a regex such as .*\w{\s\S]{100} that has in the order of 100 NFA states but the 
DFA states correspond to subsets of those states. When the total size of all those subsets grows too large, i.e, SymbolicRegexNode count reaches the limit, a switch to NFA mode takes place. However, if there are a large number NFA states but never any large subsets of those are created in the DFA construction (i.e. poweset blowup never occurs) then the thershold is not reached an all is fine.

Reply:
Using this limit therefore also makes the limit restriction on the approximated NFA size based on regex (that for example currently prohibits use of large counters as the size approximation considers counters as if they would be unwound e.g. for a regex (a{50}){20} it would be 20*50), becomes much less critical.